text/0005-is-operator.md

```diff
@@ -0,0 +1,176 @@
+
+## Best practice: Using "typed" actions
+
+From the discussion on <https://github.com/cedar-policy/cedar/issues/94>:
+
+> Guidance we have given other customers is that actions should be "typed," in the sense that they include in their name the type of the resource they apply to. So instead of writing
+>
+>     permit(
+>       principal == User::"admin",
+>       action == Action::"view",
+>       resource is Photo
+>     );
+>
+> you would write
+>
+>     permit(
+>       principal == User::"admin",
+>       action == Action::"viewPhoto",
+>       resource
+>     );
+>
+> In your schema, you indicate that the "viewPhoto" action should always be accompanied by a `Photo` for the resource, so the policy doesn't need to say that explicitly (and doesn't need to perform a runtime check).
+
+If we support an `is` operator, then customers may be encouraged to use a single "view" action for multiple resource types, rather than a new "viewX" actions for each resource type X.
+
+# Alternatives
+
+1. For the most part, users can currently solve the problems that `is` solves by specifying in the schema what types of principals and resources an action applies to, or by using `has` checks, per the examples in the [Motivation](#motivation) above.
+However, in some cases, `is` may result in policies that are easier to read.
+(Aside: the issue of "input validation" brought up in [Use case 1](#use-case-1-input-validation) could also be addressed by performing validation against the schema for every input query, but that deserves its own RFC.)
+
+2. Another way to simulate `is` is to add an `entity_type` attribute to entities, and check the type of an entity using `resource.entity_type == "File"` (for example).
+However, this requires users to manually add an entity's type as an attribute, and leads to the possibility that an entity might be created with an attribute that doesn't actually match its type.
+
+3. The behavior of `is` can also be simulated using groups.
+For each entity type (e.g., `File`), a user could create a group (e.g., `Type::"File"`) and add all entities of that type to the group.
+Then to check whether an entity has a particular type, users could use `in` (e.g., `resource in Type::"File"`).
+Like alternative (2), this requires some extra work on the part of the user (setting up the memberOf relation), and does not prevent entities from being added as members of the wrong group.
+
+# Unresolved questions
+
+1. Should we allow `is` to apply to action entities? Action entities should _always_ have type `Action`, although they may have different namespaces. So `is` could still be used to distinguish between `Namespace1::Action` and `Namespace2::Action`.
+
+2. Should users be able to indicate that an entity may be of multiple types, e.g., `principal is [User, Group]`? Can this be combined with `in` (e.g., `principal is [User, Group] in Group::"orgA"`, or even `principal is [User, Group] in [Group::"orgA", Group::"orgB"]`)?
```

Comment:
Users will ask for this :). (not necessarily in favor of doing it for the initial impl though)

Reply:
This looks confusing to me. `foo in [X,Y]` looks like set inclusion, i.e., `foo` is either in `X` or `Y`. `foo is [X,Y]` seems like it might desugar to `foo is X` _and_ `foo is Y`. Analogy to `has`: If I wrote `foo has [X,Y]` would you think I'm saying it has both attributes, or just one of them?

So whether they might ask or not, I am seeing some downsides in terms of interpreting the syntax. Maybe better not to go there.

Reply:
I disagree a little.  You say

> `foo is [X,Y]`seems like it might desugar to `foo is X` _and_ `foo is Y`

but the precedent you cite (`foo in [X,Y]`) has the other behavior (implicit _or_), which is exactly the right interpretation here as well.  What's inconsistent?

This is not a statement either way of my position on the actual proposal to allow `is [X,Y]`, which I don't have a formed opinion on yet

Reply:
Latest revision includes a note saying that we will not support this syntax. (But will reconsider if/when someone asks for it.)